### ISSUE TITLE
CMake may silently ignore new platform when switching platforms

### ISSUE THREAD

**Issue by illuhad** [illuhad/hipSYCL#316]
It seems that, when switching hipSYCL platforms, the cmake integration may ignore the new platform setting. To make things work the cache has to be cleared first. This is in principle not a new issue, and also appeared with the old cmake integration. However, previously compilation would just fail in these cases (see #297).

With the new runtime however, things will silently "just work", except that you are not on the backend that you think you are on.

This is of course highly dangerous 
* for any kind of benchmarking or production code where performance matters
* for our own development workflow - since the unit tests now also use the cmake integration, they are also affected which might cause us to draw wrong conclusions about which backends have (or don't have) issues.

This should be tackled asap. Could you take a look @fknorr @psalz ?

**Comment by fknorr** [comment]
The problem is that CMake does not recognize the `RULE_LAUNCH_COMPILE` string, which contains the platform argument, as a dependency. I'm working on it.

**Comment by illuhad** [comment]
Maybe I'm doing something wrong, but I seem to still be hitting the problem. I can reproduce e.g. by the following:
```
cmake -DCMAKE_PREFIX_PATH=`pwd`/../install/lib/cmake/hipSYCL/ -DHIPSYCL_PLATFORM=rocm -DHIPSYCL_GPU_ARCH=gfx906  $TESTS_SRC
# From observing the dryrun we can see that it correctly compiles for rocm as expected
HIPSYCL_DRYRUN=1 make
make clean
cmake -DCMAKE_PREFIX_PATH=`pwd`/../install/lib/cmake/hipSYCL/ -DHIPSYCL_PLATFORM=cpu  -DHIPSYCL_CPU_CXX=clang++ $TESTS_SRC
HIPSYCL_DRYRUN=1 make
```
In that case it errors with `syclcc fatal error: Required command line argument --hipsycl-gpu-arch or environment variable HIPSYCL_GPU_ARCH not specified` because it is still compiling for GPU. If we go the other direction, i.e. switch from CPU to GPU, it compiles silently still for CPU.

**Comment by fknorr** [comment]
I can't reproduce this with neither Ninja nor Makefiles. Try removing all build an install directories, re-install hipSYCL and build again. After that, changing the platform should trigger a recompile.

**Comment by illuhad** [comment]
Didn't help unfortunately. I can reproduce this both on my Arch Linux machine as well as on my Ubuntu 19.10 machine.

**Comment by sbalint98** [comment]
I could reproduce the problem @illuhad faced, I submitted a PR with a solution that seems to work for me.

**Comment by fknorr** [comment]
I still can't reproduce this with CMake 3.18.3 and the hipSYCL test subproject. Maybe it's down to a version requirement or policy change in your project including hipSYCL.

**Comment by illuhad** [comment]
Fixed by PR #319
